subsys/mgmt/mcumgr/smp_bt.c

```diff
@@ -103,48 +104,78 @@ static struct conn_param_data *get_conn_param_data(const struct bt_conn *conn)
 }
 
 /* Sets connection parameters for a given conn. */
-static void conn_params_set(struct bt_conn *conn, bool smp_params)
+static bool conn_params_set(struct bt_conn *conn, struct bt_le_conn_param *params)
 {
 	int ret = 0;
-	const struct bt_le_conn_param *params;
 	struct conn_param_data *cpd;
 
 	cpd = get_conn_param_data(conn);
-	params = smp_params ? SMP_BT_CONN_PARAMS : DEFAULT_CONN_PARAMS;
 
 	ret = bt_conn_le_param_update(conn, params);
 	if (ret && (ret != -EALREADY)) {
-		if (!smp_params) {
-			/* Try again to avoid stucking in SMP connection parameters. */
-			(void)k_work_reschedule(&cpd->dwork, K_MSEC(RESTORE_RETRY_TIME));
-		}
+		/* Try again to avoid stucking with incorrect connection parameters. */
+		(void)k_work_reschedule(&cpd->ework, K_MSEC(RESTORE_RETRY_TIME));
+
+		return false;
+	} else {
+		(void)k_work_cancel_delayable(&cpd->ework);
+
+		return true;
 	}
 }
 
 
 /* Work handler function for restoring the default connection parameters for the connection. */
-static void pref_conn_params_restore(struct k_work *work)
+static void on_pref_conn_params_restore(struct k_work *work)
 {
 	struct conn_param_data *cpd;
 
 	cpd = CONTAINER_OF(work, struct conn_param_data, dwork);
+	if (conn_params_set(cpd->conn, DEFAULT_CONN_PARAMS)) {
```

Comment:
Maybe we could clear the flag anyway. In case of problems the error work would still be running periodically in the background, so it would send the request